### PR TITLE
EI backend scaling and failover

### PR DESCRIPTION
**Pull request by Christoffer-Cortes**

### CODE REVIEW

File: src/functionaltests/java/com/ericsson/ei/scaling/ScalingAndFailoverSteps.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        System.setOut(printStream);
+    }
+
+    @After("@ScalingAndFailoverScenario")
+    public void afterScenario() {
+        printStream.close();
+    }
+
+    @Given("^\"([0-9]+)\" additional instance(s)? of Eiffel Intelligence$")
+    public void additional_eiffel_intelligence_instances(int multiple, String plural) throws Exception {
```
Comment (evasiba-ericsson) [inline src/functionaltests/java/com/ericsson/ei/scaling/ScalingAndFailoverSteps.java]: this is to big. split it in smaller methods.

File: src/functionaltests/resources/features/scalingAndFailover.feature [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+@ScalingAndFailoverFeature
+Feature: Scaling and Failover test
+
+  @ScalingAndFailoverScenario
+  Scenario: Scaling and Failover
+    Given "3" additional instances of Eiffel Intelligence
+    When "1000" eiffel events are sent
+    And additional instances are closed
+    Then all event messages are processed
+    And unprocessed events are affected by failover
```
Comment (evasiba-ericsson) [inline src/functionaltests/resources/features/scalingAndFailover.feature]: the name of the step is misleading. we want to prove the unprocessed event are not being affected by instance failures.

File: src/functionaltests/java/com/ericsson/ei/scaling/ScalingAndFailoverSteps.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+@TestPropertySource(properties = { "logging.level.com.ericsson.ei.handlers.EventHandler=INFO" })
+@Ignore
+@AutoConfigureMockMvc
+public class ScalingAndFailoverSteps extends FunctionalTestBase {
+    private static final String EVENT_DUMMY = "src/functionaltests/resources/scale_and_failover_dummy.json";
+
+    @LocalServerPort
+    private int port;
+    private List<Integer> portList = new ArrayList<Integer>();
```
Comment (ned29) [inline src/functionaltests/java/com/ericsson/ei/scaling/ScalingAndFailoverSteps.java]: Diamonds type in new ArrayList< Integer > unneeded

File: src/functionaltests/java/com/ericsson/ei/scaling/ScalingAndFailoverSteps.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+    private List<WebApplicationContext> appContextList = new ArrayList<WebApplicationContext>();
```
Comment (ned29) [inline src/functionaltests/java/com/ericsson/ei/scaling/ScalingAndFailoverSteps.java]: the same

File: src/functionaltests/java/com/ericsson/ei/scaling/ScalingAndFailoverSteps.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+    @Autowired
+    private MockMvc mockMvc;
+    private List<MockMvc> mockMvcList = new ArrayList<MockMvc>();
```
Comment (ned29) [inline src/functionaltests/java/com/ericsson/ei/scaling/ScalingAndFailoverSteps.java]: the same

File: src/functionaltests/java/com/ericsson/ei/scaling/ScalingAndFailoverSteps.java [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+
+    private List<String> eventsIdList = new ArrayList<String>();
```
Comment (ned29) [inline src/functionaltests/java/com/ericsson/ei/scaling/ScalingAndFailoverSteps.java]: the same

File: src/functionaltests/java/com/ericsson/ei/scaling/ScalingAndFailoverSteps.java [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+    public void messages_processed() throws Exception {
+        List<String> missingEventIds = verifyEventsInDB(eventsIdList);
+        LOGGER.debug("Missing events: {}", missingEventIds.toString());
+        assertEquals("Number of events missing in DB: " + missingEventIds.size(), 0, missingEventIds.size());
+    }
+
+    @Then("^event messages are unaffected by instance failure$")
+    public void events_failover() throws Exception {
+        List<Integer> receivedCount = new ArrayList<Integer>();
+        List<Integer> processedCount = new ArrayList<Integer>();
```
Comment (ned29) [inline src/functionaltests/java/com/ericsson/ei/scaling/ScalingAndFailoverSteps.java]: the same

File: src/functionaltests/java/com/ericsson/ei/scaling/ScalingAndFailoverSteps.java [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+    /**
+     * Counts events received or processed by EventHandler
+     *
+     * @param type
+     *            either received or processed
+     * @return list of event occurrences
+     */
+    private List<Integer> eventMessageCounter(String type) {
+        List<Integer> list = new ArrayList<Integer>();
```
Comment (ned29) [inline src/functionaltests/java/com/ericsson/ei/scaling/ScalingAndFailoverSteps.java]: the same

File: src/functionaltests/java/com/ericsson/ei/scaling/TestScalingAndFailoverRunner.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+import org.junit.runner.RunWith;
+
+import cucumber.api.CucumberOptions;
+import cucumber.api.junit.Cucumber;
+
+@RunWith(Cucumber.class)
+@CucumberOptions(features = "src/functionaltests/resources/features/scalingAndFailover.feature", glue = {
+        "com.ericsson.ei.scaling" }, plugin = {
+                "html:target/cucumber-reports/TestSubscriptionTriggerRunner" }, monochrome = false)
```
Comment (ned29) [inline src/functionaltests/java/com/ericsson/ei/scaling/TestScalingAndFailoverRunner.java]: monochrome = false, can be deleted in CucumberOptions annotation the monochrome set false as default so unneeded here

File: src/functionaltests/java/com/ericsson/ei/utils/MultiOutputStream.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
+package com.ericsson.ei.utils;
+
+import java.io.IOException;
+import java.io.OutputStream;
+
+public class MultiOutputStream extends OutputStream
+{
+    OutputStream[] outputStreams;
```
Comment (ned29) [inline src/functionaltests/java/com/ericsson/ei/utils/MultiOutputStream.java]: I thinks should be private
and general comment about class move the "{" up in all your methods, should be after method not from the new line

File: src/main/java/com/ericsson/ei/handlers/EventHandler.java [hunk 11]
```diff
@@ -43,35 +46,34 @@
...
 
     @Async
     public void onMessage(Message message, Channel channel) throws Exception {
-        byte[] messageBody = message.getBody();
+        String messageBody = new String(message.getBody());
+        ObjectMapper objectMapper = new ObjectMapper();
+        JsonNode node = objectMapper.readTree(messageBody);
+        String id = node.get("meta").get("id").toString();
+        String port = environment.getProperty("local.server.port");
+        log.info("Event {} received on port {}", id, port);
```
Comment (ned29) [inline src/main/java/com/ericsson/ei/handlers/EventHandler.java]: set level to debug

File: src/main/java/com/ericsson/ei/handlers/EventHandler.java [hunk 12]
```diff
@@ -43,35 +46,34 @@
...
+
         eventReceived(messageBody);
         long deliveryTag = message.getMessageProperties().getDeliveryTag();
         channel.basicAck(deliveryTag, false);
+
+        log.info("Event {} processed on port {}", id, port);
```
Comment (ned29) [inline src/main/java/com/ericsson/ei/handlers/EventHandler.java]: to debug

**Pull request merged**